sdk/containerregistry/container-registry/src/blob/containerRegistryBlobClient.ts

```diff
@@ -0,0 +1,400 @@
+   *
+   * @param manifest - the manifest to upload. If a resettable stream (a factory function that returns a stream) is provided, it may be called multiple times. Each time the function is called, a fresh stream should be returned.
+   */
+  public async uploadManifest(
+    manifest: (() => NodeJS.ReadableStream) | NodeJS.ReadableStream | OciManifest,
+    options?: UploadManifestOptions
+  ): Promise<UploadManifestResult> {
+    const { span, updatedOptions } = createSpan(
+      "ContainerRegistryBlobClient-uploadManifest",
+      options
+    );
+
+    try {
+      let manifestBody: Buffer | NodeJS.ReadableStream;
+
+      if (isReadableStream(manifest)) {
+        manifestBody = await readStreamToEnd(manifest);
+      } else if (typeof manifest === "function") {
+        manifestBody = await readStreamToEnd(manifest());
+      } else {
+        const serialized = serializer.serialize(Mappers.OCIManifest, manifest);
+        manifestBody = Buffer.from(JSON.stringify(serialized));
+      }
+
+      const tagOrDigest = options?.tag ?? (await calculateDigest(manifestBody));
+
+      const { dockerContentDigest } = await this.client.containerRegistry.createManifest(
+        this.repositoryName,
+        tagOrDigest,
+        manifestBody,
+        { contentType: KnownManifestMediaType.OciManifestMediaType, ...updatedOptions }
+      );
+
+      if (!dockerContentDigest) {
+        throw new Error("Digest not provided");
+      }
+
+      return { digest: dockerContentDigest };
+    } catch (e) {
+      span.setStatus({ code: SpanStatusCode.ERROR, message: e.message });
+      throw e;
+    } finally {
+      span.end();
+    }
+  }
+
+  /**
+   * Downloads the manifest for an OCI artifact
+   *
+   * @param tagOrDigest - a tag or digest that identifies the artifact
+   * @returns - the downloaded manifest
+   */
+  public async downloadManifest(
+    tagOrDigest: string,
+    options?: DownloadManifestOptions
+  ): Promise<DownloadManifestResult> {
+    const { span, updatedOptions } = createSpan(
+      "ContainerRegistryBlobClient-downloadManifest",
+      options
+    );
+
+    try {
+      const rawResponse = await new Promise<FullOperationResponse>(async (resolve) => {
+        this.client.containerRegistry.getManifest(this.repositoryName, tagOrDigest, {
+          accept: KnownManifestMediaType.OciManifestMediaType,
+          ...updatedOptions,
+          onResponse: (rawResponseProvided, flatResponse, error) => {
+            updatedOptions?.onResponse?.(rawResponseProvided, flatResponse, error);
+            resolve(rawResponseProvided);
+          },
+        });
+      });
+
+      const digest = rawResponse.headers.get("Docker-Content-Digest");
+      if (!digest) {
+        throw new Error("Expected Docker-Content-Digest header in response");
+      }
+
+      const body = rawResponse.bodyAsText;
+      if (!body) {
+        throw new Error("downloadManifest did not return a text body");
+      }
```

Comment:
all of this feels very odd, like we're second-guessing what the server might return. If we do this, I think we should be throwing a RestError with the server's raw response so it's easier to handle and figure out that this is a service issue.

Reply:
Really, most of this is to satisfy the type checker. A RestError makes sense though.

Reply:
don't get me wrong, I'm not a fan of this logic either, but couldn't think of a better way to do it. I wonder if I can do a Swagger transform so we get everything we need out without having to resort to this? I'll have a play around. We'd still have to keep it as text and deserialize ourselves so we can do the digest comparison.

Reply:
Good news, I think that I managed to take care of this in the Swagger. I like this a lot more.